### PR TITLE
libpmem fixes

### PR DESCRIPTION
**Pull request by lukaszstolarczuk**
PR created for review purpose

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/fio/121)

### CODE REVIEW

File: engines/libpmem.c [hunk 1]
```diff
@@ -226,10 +211,11 @@ static int fio_libpmem_close_file(struct thread_data *td, struct fio_file *f)
...
-	dprint(FD_IO,"DEBUG fio_libpmem_close_file\n");
-	dprint(FD_IO,"td->o.odirect %d \n",td->o.odirect);
+	dprint(FD_IO, "DEBUG fio_libpmem_close_file\n");
+	dprint(FD_IO, "td->o.odirect %d td->o.sync_io %d\n",
+					td->o.odirect, td->o.sync_io);
 
-	if (!td->o.odirect) {
+	if (!td->o.odirect || !td->o.sync_io) {
 		dprint(FD_IO,"pmem_drain\n");
 		pmem_drain();
```
Comment (lukaszstolarczuk) [inline engines/libpmem.c:220]: shouldn't we call here `pmem_persist`?

**Review by grom72 — COMMENTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/fio/121)* status: all files reviewed, 2 unresolved discussions (waiting on @lukaszstolarczuk)

___
*[examples/libpmem.fio, line 52 at r1](https://reviewable.io/reviews/pmem/fio/121#-MR3n5suCdZIRhi0RHPz:-MR3n5sv9ui26c7Ez2tw:b-se47c3) ([raw file](https://github.com/pmem/fio/blob/3fe4a137179f3c39c25c163a7d2d88c5b9fbac1c/examples/libpmem.fio#L52)):*
> ```
> #filename=/mnt/pmem/somefile
> ```
does it work also with dev-dax?

**Comment by lukaszstolarczuk** [comment]
> > Reviewed 2 of 2 files at r1.
> > _[Reviewable](https://reviewable.io/reviews/pmem/fio/121)_ status: all files reviewed, 2 unresolved discussions (waiting on @lukaszstolarczuk)
> 
> _[examples/libpmem.fio, line 52 at r1](https://reviewable.io/reviews/pmem/fio/121#-MR3n5suCdZIRhi0RHPz:-MR3n5sv9ui26c7Ez2tw:b-se47c3) ([raw file](https://github.com/pmem/fio/blob/3fe4a137179f3c39c25c163a7d2d88c5b9fbac1c/examples/libpmem.fio#L52)):_
> 
> > ```
> > #filename=/mnt/pmem/somefile
> > ```
> 
> does it work also with dev-dax?

It suppose to - this is some of the next steps to confirm it, fix if needed, and get rid (mark depreciated) of `dev-dax` engine.
I'm not sure if all of the internals are setup correctly for devdax, we'll check it.

Why you're asking? Would you (or a client) be interested in using it that way?

**Review by grom72 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/fio/121)* status: all files reviewed, 2 unresolved discussions (waiting on @lukaszstolarczuk)

___
*[examples/libpmem.fio, line 52 at r1](https://reviewable.io/reviews/pmem/fio/121#-MR3n5suCdZIRhi0RHPz:-MR4mMiJDcWpAZ-9-w_S:b-wmyg71) ([raw file](https://github.com/pmem/fio/blob/3fe4a137179f3c39c25c163a7d2d88c5b9fbac1c/examples/libpmem.fio#L52)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

It suppose to work, but this is some of the next steps: to confirm it, fix if needed, and get rid (mark depreciated) of `dev-dax` engine. I'm not sure if all of the internals are setup correctly for devdax, we'll check it.

Why you're asking? Would you (or a client) be interested in using it that way? 
</blockquote></details>

yes, with librpma we use dev-dax often.

File: engines/libpmem.c (hunk 1 above)
Comment (pbalcer) [reply engines/libpmem.c:220]: No, the !sync_io condition means that the memcpy flag is NODRAIN.

I think the the direct flag should be removed from this condition - non-temporals and temporal both call a drain.

Comment (lukaszstolarczuk) [reply engines/libpmem.c:220]: yeah, right, direct should be removed from this equation at all.

We're talking about the close_file function, which IMHO should call drain/persist if it was not called previously (and this is the case if all of the IO was not sync).... or am I wrong and if user wants NO_DRAIN then NO_DRAIN it is, in the whole FIO workflow...?
Not that this is important at this point... maybe we should just remove this drain here...?

**Comment by lukaszstolarczuk** [comment]
> > _[Reviewable](https://reviewable.io/reviews/pmem/fio/121)_ status: all files reviewed, 2 unresolved discussions (waiting on @lukaszstolarczuk)
> 
> _[examples/libpmem.fio, line 52 at r1](https://reviewable.io/reviews/pmem/fio/121#-MR3n5suCdZIRhi0RHPz:-MR4mMiJDcWpAZ-9-w_S:b-wmyg71) ([raw file](https://github.com/pmem/fio/blob/3fe4a137179f3c39c25c163a7d2d88c5b9fbac1c/examples/libpmem.fio#L52)):_
> 
> Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…
> yes, with librpma we use dev-dax often.

use it to...?

and does this mean you want to stick with dev-dax engine or you want to try to use libpmem engine?

**Review by grom72 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/fio/121)* status: all files reviewed, 2 unresolved discussions (waiting on @lukaszstolarczuk and @pbalcer)

___
*[examples/libpmem.fio, line 52 at r1](https://reviewable.io/reviews/pmem/fio/121#-MR3n5suCdZIRhi0RHPz:-MR9iB9w77Dhl0aACyNg:b-vq6aln) ([raw file](https://github.com/pmem/fio/blob/3fe4a137179f3c39c25c163a7d2d88c5b9fbac1c/examples/libpmem.fio#L52)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

use it to...?

and does this mean you want to stick with dev-dax engine or you want to try to use libpmem engine? 
</blockquote></details>

I would like to use libpmem engine with dev-dax

File: engines/libpmem.c (hunk 1 above)
Comment (lukaszstolarczuk) [reply engines/libpmem.c]: I've removed the drain from the close function at all, seems not necessary.

**Comment by lukaszstolarczuk** [comment]
PR on upstream created: https://github.com/axboe/fio/pull/1185

**Pull request closed without merging**